imap/http_dav.c

```diff
@@ -4974,33 +4976,52 @@ int meth_delete(struct transaction_t *txn, void *params)
     if (!txn->req_tgt.resource) {
         /* DELETE collection */
 
+        /* Open mailbox for reading */
+        r = mailbox_open_irl(txn->req_tgt.mbentry->name, &mailbox);
+        if (r) {
+            syslog(LOG_ERR, "http_mailbox_open(%s) failed: %s",
+                    txn->req_tgt.mbentry->name, error_message(r));
+            txn->error.desc = error_message(r);
+            ret = HTTP_SERVER_ERROR;
+            goto done;
+        }
+
+        /* Check any preconditions */
+        ret = dparams->check_precond(txn, params, mailbox, NULL, NULL, 0);
+
+        switch (ret) {
+            case HTTP_OK:
+                break;
+            case HTTP_LOCKED:
+                txn->error.precond = DAV_NEED_LOCK_TOKEN;
+                txn->error.resource = txn->req_tgt.path;
+                GCC_FALLTHROUGH
+            default:
+                /* We failed a precondition - don't perform the request */
+                r = ret ? ret : HTTP_SERVER_ERROR;
+                goto done;
+        }
+
         if (dparams->delete) {
             /* Do special processing on all resources */
             struct delete_rock drock = { txn, NULL, dparams->delete };
 
-            /* Open mailbox for reading */
-            r = mailbox_open_irl(txn->req_tgt.mbentry->name, &mailbox);
-            if (r) {
-                syslog(LOG_ERR, "http_mailbox_open(%s) failed: %s",
-                       txn->req_tgt.mbentry->name, error_message(r));
-                txn->error.desc = error_message(r);
-                return HTTP_SERVER_ERROR;
-            }
-
             /* Open the DAV DB corresponding to the mailbox */
```

Comment:
As per my comment above, we would NOT make the change above and we would do the following before opening DAV DB:

ret = dparams->delete(txn, mailbox, NULL, NULL);
if (ret) goto done;